integration_tests/src/lib.rs

```diff
@@ -2080,56 +2085,205 @@ mod test {
                 )),
             };
 
-        // XXX Crucible uses std::sync::mpsc::Receiver, not
-        // tokio::sync::mpsc::Receiver, so use tokio::task::block_in_place here.
-        // Remove that when Crucible changes over to the tokio mpsc.
-        let volume =
-            tokio::task::block_in_place(|| Volume::construct(vcr, None))?;
-        volume.activate(0)?;
+        let volume = Volume::construct(vcr, None).await?;
+        volume.activate(0).await?;
 
         // Validate that source blocks originally come from the read-only parent
         {
-            let buffer = Buffer::new(volume.total_size()? as usize);
-            tokio::task::block_in_place(|| {
-                volume.read(
+            let buffer = Buffer::new(volume.total_size().await? as usize);
+            volume
+                .read(
                     Block::new(0, BLOCK_SIZE.trailing_zeros()),
                     buffer.clone(),
                 )
-            })?
-            .block_wait()?;
+                .await?
+                .wait()
+                .await?;
 
-            assert_eq!(*buffer.as_vec(), random_buffer);
+            assert_eq!(*buffer.as_vec().await, random_buffer);
         }
 
         // Validate a flush works
-        volume.flush(None)?.block_wait()?;
+        volume.flush(None).await?.wait().await?;
 
         // Write one block of 0x00 in, validate with a read
         volume
             .write(
                 Block::new(0, BLOCK_SIZE.trailing_zeros()),
                 Bytes::from(vec![0u8; BLOCK_SIZE]),
-            )?
-            .block_wait()?;
+            )
+            .await?
+            .wait()
+            .await?;
 
         {
-            let buffer = Buffer::new(volume.total_size()? as usize);
-            tokio::task::block_in_place(|| {
-                volume.read(
+            let buffer = Buffer::new(volume.total_size().await? as usize);
+            volume
+                .read(
                     Block::new(0, BLOCK_SIZE.trailing_zeros()),
                     buffer.clone(),
                 )
-            })?
-            .block_wait()?;
+                .await?
+                .wait()
+                .await?;
 
-            let buffer_vec = buffer.as_vec();
+            let buffer_vec = buffer.as_vec().await;
 
             assert_eq!(buffer_vec[..BLOCK_SIZE], vec![0u8; BLOCK_SIZE]);
             assert_eq!(buffer_vec[BLOCK_SIZE..], random_buffer[BLOCK_SIZE..]);
         }
 
         // Validate a flush still works
-        volume.flush(None)?.block_wait()?;
+        volume.flush(None).await?.wait().await?;
+
+        Ok(())
+    }
+
+    #[tokio::test(flavor = "multi_thread", worker_threads = 8)]
+    async fn integration_test_wait_all_is_ordered_for_writes() -> Result<()> {
+        // In a loop, test that submitting a big batch of operations and waiting
+        // on their BlockReqWaiters in parallel is correct. It's important that
+        // this test use multiple worker_threads!
```

Comment:
To be clear I understand, this test is submitting overlapping writes, but making sure that they
are performed in the order they were submitted?

Reply:
Correct.